google-cloud-firestore/lib/google/cloud/firestore/query_listener.rb

```diff
@@ -0,0 +1,566 @@
+          sleep @backoff[:delay]
+
+          retry
+        rescue RestartStream
+          retry
+        rescue StandardError
+          raise Google::Cloud::Error.from_error(e)
+        end
+
+        # rubocop:enable all
+
+        # @private Collects changes and produces a QuerySnapshot.
+        # Uses RBTree to hold a sorted list of DocumentSnapshot objects and to
+        # make inserting and removing objects much more efficent.
+        class Inventory
+          attr_accessor :current
+          attr_reader :resume_token, :read_time
+
+          def initialize query
+            @query = query
+            @pending = {
+              add: [],
+              delete: []
+            }
+            @current = nil
+            @resume_token = nil
+            @read_time = nil
+            @tree = RBTree.new
+            @tree.readjust(&method(:query_comparison_proc))
+            @old_order = nil
+          end
+
+          def current?
+            @current
+          end
+
+          def add doc_grpc
+            @pending[:add] << doc_grpc
+          end
+
+          def delete doc_path
+            @pending[:delete] << doc_path
+          end
+
+          def pending?
+            @pending[:add].any? || @pending[:delete].any?
+          end
+
+          def clear_pending
+            @pending[:add].clear
+            @pending[:delete].clear
+          end
+
+          def size
+            @tree.size
+          end
+
+          def reset
+            # clears all but query, resume token, read time, and old order
+            clear_pending
+
+            @current = nil
+
+            @tree.clear
+          end
+
+          def persist resume_token, read_time
+            # Remove the deleted documents
+            @pending[:delete].each do |doc_path|
+              remove_doc_from_tree doc_path
+            end
+
+            # Add/update the changed documents
+            @pending[:add].each do |doc_grpc|
+              removed_doc = remove_doc_from_tree doc_grpc.name
+              added_doc = DocumentSnapshot.from_document(
+                doc_grpc, @query.client, read_at: read_time
+              )
+
+              if removed_doc && removed_doc.updated_at >= added_doc.updated_at
+                # Restore the removed doc if the added doc isn't newer
+                added_doc = removed_doc
+              end
+
+              add_doc_to_tree added_doc
+            end
+
+            @resume_token = resume_token
+            @read_time = read_time
+            clear_pending
+          end
+
+          def changes?
+            # Act like there are changes if we have never run before
+            return true if @old_order.nil?
+            added_paths, deleted_paths, changed_paths = \
+              change_paths current_order, @old_order
+            added_paths.any? || deleted_paths.any? || changed_paths.any?
+          end
+
+          def current_docs
+            @tree.keys
```

Comment:
Does this create a copy? You need to ensure that changes to the `inventory` don't affect QuerySnapshots that have already been raised.

Reply:
No, the red-black tree library we are using does not create a copy. Instead, we are storing the ordered document paths and update time in a hash. We have an old hash from the previous callback, and a new hash for the current callback. We calculate which docs have been added, removed, and updated (including their indexes) using these hashes.

Changing a document snapshot in inventory removes the old document snapshot object from the tree and then adds the new document snapshot object to the tree. A query snapshot that has been passed to the callback will not be affected, because the document snapshot objects are immutable.

Target changes are pulled from the stream in order, and are handled sequentially. The query snapshot will be sent to the callback (raised) before another target change is handled.

Reply:
Yes, each query snapshot and document snapshot are considered immutable. Previous snapshots and future snapshots won't change the current snapshot contents.